### PR TITLE
Override `optimizer_zero_grad` when using the `IPUStrategy`

### LINKED ISSUE #12903: Additional `zero_grad` and `backward` being performed when using IPU
## 🐛 Bug

When training on the IPU, the zero gradients and backward pass are compiled into the IPU program. This means that instead of: 

```python
for batch, target in training_data:
    # Zero gradients
    optimizer.zero_grad()

    # Run model.
    _, loss = model(batch, target)

    # Back propagate the gradients.
    loss.backward()

    # Update the weights.
    optimizer.step()
```
You would use:
```python
for batch, target in training_data:
    # Performs forward pass, loss function evaluation,
    # backward pass and weight update in one go on the device.
    _, loss = poptorch_model(batch, target)
```

However, when [closure()](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loops/optimization/optimizer_loop.py#L131-L143) is executed in [optimizer_step()](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/plugins/precision/ipu.py#L54-L78), both `self._zero_grad_fn()` and `self._backward_fn(step_output.closure_loss)` are still being executed. This is leading to duplicate work and decreased throughput performance.

### To Reproduce

<details>
<summary>
repro.py
</summary>

```python
import torch
import poptorch
import pytorch_lightning as pl
import torch.nn.functional as F
from torchvision import datasets, models
from torchvision.transforms import ToTensor
from pytorch_lightning.strategies.ipu import IPUStrategy


class Model(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.model = models.resnet18(pretrained=False)

    def forward(self, x):
        return self.model(x)

    def training_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self.model(x)
        loss = F.cross_entropy(y_hat, y)
        return loss

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=1e-3)


def main():
    batch_size = 32
    # data
    mnist = datasets.FakeData(
        size=100 * batch_size,
        image_size=(3, 28, 28),
        num_classes=10,
        transform=ToTensor(),
    )
    opts = poptorch.Options()
    train_loader = poptorch.DataLoader(
        options=opts,
        dataset=mnist,
        batch_size=batch_size,
        drop_last=True,
        mode=poptorch.DataLoaderMode.Async,
    )
    # model
    model = Model()
    # training
    trainer = pl.Trainer(
        accelerator="ipu",
        precision=16,
        max_epochs=1,
        strategy=IPUStrategy(training_opts=opts),
    )
    trainer.fit(model, train_loader)


if __name__ == "__main__":
    main()
```
</details>

### Expected behavior

`self._zero_grad_fn()` and `self._backward_fn(step_output.closure_loss)` should not be executed in `closure()`.

### Environment

* Packages:
  - poptorch
  - pytorch-lightning==1.6.1
  - torch @ https://download.pytorch.org/whl/cpu/torch-1.10.0%2Bcpu-cp37-cp37m-linux_x86_64.whl
  - torchvision==0.11.1
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.7.5
        - version:           18.04.1-Ubuntu SMP Wed Sep 29 10:59:42 UTC 2021

### Additional context

In the following profile (top: PyTorch , bottom: PyTorch Lightning) the `copyWeightsToHost` are from the additional zero gradients, and I believe (but have not confirmed) that the extra execution time is due to the additional backward pass.

image

cc @SeanNaren

### PR DESCRIPTION
**Pull request by HMellor**
## What does this PR do?

Overrides the `optimizer_zero_grad` function to do nothing when `IPUStrategy` is used. This is because this operation is handled internally by the IPU.

Fixes #12903

### Does your PR introduce any breaking changes? If yes, please list them.

None.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/strategies/ipu.py [hunk 1]
```diff
@@ -260,6 +266,18 @@ def to_tensor(x):
         args = apply_to_collection(args, dtype=(int, float), function=to_tensor)
         return args
 
+    def _disable_zero_grad(self):
```
Comment (SeanNaren) [inline pytorch_lightning/strategies/ipu.py:269]:
```suggestion
    def _disable_zero_grad(self) -> None:
```

**Review by SeanNaren — APPROVED** [review]

**Comment by SeanNaren** [comment]
Great first PR @HMellor was awesome working with you on this :)

**Review by kaushikb11 — APPROVED** [review]

File: pytorch_lightning/strategies/ipu.py [hunk 2]
```diff
@@ -260,6 +266,18 @@ def to_tensor(x):
...
+        if is_overridden("optimizer_zero_grad", self.lightning_module):
+            warning_cache.warn(
+                "You have overridden the `LightningModule.optimizer_zero_grad` hook but it will be ignored since"
+                " IPUs handle the zeroing of gradients internally."
+            )
+
+        def override_fn(*args: Any, **kwargs: Any) -> None:
+            pass
+
+        self.lightning_module.optimizer_zero_grad = override_fn  # type: ignore[assignment]
```
Comment (carmocca) [inline pytorch_lightning/strategies/ipu.py]: Same comment as https://github.com/PyTorchLightning/pytorch-lightning/pull/12842#discussion_r860843496

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/strategies/ipu.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/strategies/ipu.py]: Pushed the change in https://github.com/PyTorchLightning/pytorch-lightning/pull/12913/commits/80f092070d45742a024ac893539ac09a699c6958

File: CHANGELOG.md [hunk 3]
```diff
@@ -147,6 +147,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed an issue to ensure all the checkpoint states are saved in a common filepath with `DeepspeedStrategy` ([#12887](https://github.com/PyTorchLightning/pytorch-lightning/pull/12887))
 
 
+- Stopped `optimizer_zero_grad` from being called after IPU execution ([#12903](https://github.com/PyTorchLightning/pytorch-lightning/issues/12903))
```
Comment (rohitgr7) [inline CHANGELOG.md:150]:
```suggestion
- Stopped `optimizer_zero_grad` from being called after IPU execution ([#12913](https://github.com/PyTorchLightning/pytorch-lightning/issues/12913))
```

File: pytorch_lightning/strategies/ipu.py [hunk 4]
```diff
@@ -260,6 +263,16 @@ def to_tensor(x):
...
 
+    def _disable_zero_grad(self) -> None:
+        lightning_module = self.lightning_module
+        if is_overridden("optimizer_zero_grad", lightning_module):
+            assert lightning_module is not None  # `is_overridden` returns False otherwise
+            rank_zero_warn(
+                "You have overridden the `LightningModule.optimizer_zero_grad` hook but it will be ignored since"
+                " IPUs handle the zeroing of gradients internally."
+            )
+            lightning_module.optimizer_zero_grad = None  # type: ignore[assignment]
```
Comment (ananthsub) [inline pytorch_lightning/strategies/ipu.py:274]: This should be restored at the end of execution to avoid patching tbh module with visible side effects to users

Comment (HMellor) [reply pytorch_lightning/strategies/ipu.py:274]: Would this happen in the `teardown` function? Similar to how the PopTorch `DataLoader` is handled?

Comment (carmocca) [reply pytorch_lightning/strategies/ipu.py:274]: Yes

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/strategies/ipu.py (hunk 4 above)
Comment (HMellor) [reply pytorch_lightning/strategies/ipu.py:274]: Ok, I'll implement that. 

@carmocca, `lightning_module.optimizer_zero_grad = None` can't be inside the `if is_overriden` block like it is in the commit. Otherwise the only way the IPU will be executed correctly is if the user overrides `optimizer_zero_grad` themselves.

File: pytorch_lightning/strategies/ipu.py [hunk 5]
```diff
@@ -134,6 +135,11 @@ def setup(self, trainer: "pl.Trainer") -> None:
 
         super().setup(trainer)
 
+        # disable the `optimizer_zero_grad` function by setting it to `None`.
+        # this is because the IPU zeros the gradients internally
+        self._optimizer_zero_grad_original = self.lightning_module.optimizer_zero_grad
```
Comment (carmocca) [inline pytorch_lightning/strategies/ipu.py:141]: You need to define this in the `__init__` too, as done for `_update_dataloader_original`

Should resolve the current IPU failures

**Review by carmocca — APPROVED** [review]
